planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,476 @@
+package com.onthegomap.planetiler.util;
+
+import com.onthegomap.planetiler.geo.DouglasPeuckerSimplifier;
+import com.onthegomap.planetiler.geo.GeoUtils;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import org.locationtech.jts.algorithm.Angle;
+import org.locationtech.jts.geom.Coordinate;
+import org.locationtech.jts.geom.CoordinateXY;
+import org.locationtech.jts.geom.Geometry;
+import org.locationtech.jts.geom.GeometryComponentFilter;
+import org.locationtech.jts.geom.GeometryFactory;
+import org.locationtech.jts.geom.LineString;
+import org.locationtech.jts.geom.PrecisionModel;
+
+
+public class LoopLineMerger {
+  final List<LineString> input = new ArrayList<>();
+  private final List<Node> output = new ArrayList<>();
+  int nodes = 0;
+  int edges = 0;
+  private PrecisionModel precisionModel = new PrecisionModel(GeoUtils.TILE_PRECISION);
+  private GeometryFactory factory = new GeometryFactory(precisionModel);
+  private double minLength = 0.0;
+  private double loopMinLength = 0.0;
+  private double stubMinLength = 0.0;
+  private double tolerance = 0.0;
+  private boolean mergeStrokes = false;
+
+  public LoopLineMerger setPrecisionModel(PrecisionModel precisionModel) {
+    this.precisionModel = precisionModel;
+    factory = new GeometryFactory(precisionModel);
+    return this;
+  }
+
+  public LoopLineMerger setMinLength(double minLength) {
+    this.minLength = minLength;
+    return this;
+  }
+
+  public LoopLineMerger setLoopMinLength(double loopMinLength) {
+    this.loopMinLength = loopMinLength;
+    return this;
+  }
+
+  public LoopLineMerger setStubMinLength(double stubMinLength) {
+    this.stubMinLength = stubMinLength;
+    return this;
+  }
+
+  public LoopLineMerger setTolerance(double tolerance) {
+    this.tolerance = tolerance;
+    return this;
+  }
+
+  public LoopLineMerger setMergeStrokes(boolean mergeStrokes) {
+    this.mergeStrokes = mergeStrokes;
+    return this;
+  }
+
+  public LoopLineMerger add(Geometry geometry) {
+    geometry.apply((GeometryComponentFilter) component -> {
+      if (component instanceof LineString lineString) {
+        input.add(lineString);
+      }
+    });
+    return this;
+  }
+
+  private void degreeTwoMerge() {
+    for (var node : output) {
+      degreeTwoMerge(node);
+    }
+  }
+
+  private Edge degreeTwoMerge(Node node) {
+    if (node.getEdges().size() == 2) {
+      Edge a = node.getEdges().getFirst();
+      Edge b = node.getEdges().get(1);
+      return mergeTwoEdges(node, a, b);
+    }
+    return null;
+  }
+
+  private Edge mergeTwoEdges(Node node, Edge a, Edge b) {
```

Comment:
What do you think about this small tweak to make a best-effort attempt to preserve line segment directions from the input?

```suggestion
  private Edge mergeTwoEdges(Node node, Edge edge1, Edge edge2) {
    // attempt to preserve segment directions from the original line
    // when: A << N -- B then output C reversed from B to A
    // when: A >> N -- B then output C from A to B
    Edge a = edge1.main ? edge2 : edge1;
    Edge b = edge1.main ? edge1 : edge2;
```